### DISCUSSION TITLE
Help with codec speex (error: no common audio codecs - rejected)

### DISCUSSION THREAD

**Discussion by mrMiimo** [baresip/baresip#2323] [Q&A] [answered]
Hello everybody,
I have an intercom/doorbell that uses Flexisip as a SIP server, it's pretty much closed as settings. I saw that it is setup and use the codec:
```
[audio_codec_1]
mime=speex
rate=8000
channels=1
enabled=1
```
 I can authenticate and register into the SIP server and, if I call from a linphone client, I am able to answer and speak.
Unfortunately I can't set Baresip to the same codec as the server (speex/8000/1) this is the log I get in Baresip:
```
pi@raspberrypi4:~/.baresip $ baresip 
baresip v1.0.0 Copyright (C) 2010 - 2020 Alfred E. Heggestad et al.
Local network address:  IPv4=eth0|192.168.1.28  IPv6=eth0|fdcf:4cbc:9db9:4005:ed44:52a2:6606:ddee
httpd: listening on 0.0.0.0:8000
aucodec: opus/48000/2
aucodec: AMR-WB/16000/1
aucodec: AMR/8000/1
aucodec: PCMU/8000/1
aucodec: PCMA/8000/1
aufilt: speex_pp
ausrc: alsa
auplay: alsa
medianat: stun
medianat: turn
medianat: ice
mediaenc: srtp
mediaenc: srtp-mand
mediaenc: srtp-mandf
franzoi-outlook.com-RASPI@2123921.bs.iotleg.com: Using media encryption 'srtp'
Populated 1 account
Populated 3 contacts
Populated 5 audio codecs
Populated 1 audio filter
Populated 0 video codecs
Populated 0 video filters
baresip is ready.
xxxxxxx-outlook.com-RASPI@xxxxx.xxx.com: {0/TCP/v4} 200 Registration successful (Flexisip/1.0.12 (sofia-sip-nta/2.0)) [1 binding]
All 1 useragent registered successfully! (105 ms)
ua: using AF from sdp offer: af=AF_INET
stream: audio: starting mediaenc 'srtp' (wait_secure=0)
call: no common audio codecs - rejected
```
Is there a way to be able to set or install this codec (speex) in Baresip?
Thanks in advance. Cheers,
R.

**Comment by whyteks** [comment]
Have you got

**module                             speex.so** 

in your config? I don't think so, as if you did, you would see:


**aucodec: speex/32000/2
aucodec: speex/16000/2
aucodec: speex/8000/2
aucodec: speex/32000/1
aucodec: speex/16000/1
aucodec: speex/8000/1**

on startup.

**Comment by mrMiimo** [comment]
Exact!
the speex.so module is not installed.

it is only mentioned in the configuration file, but I have inspected the directory `/usr/lib/baresip/modules/` and there is no

**Is it possible to install the module?** 
I searched around but i can't find info and unfortunately i'm tied to this codec.

**Comment by cspiel1** [comment]
The speex codec is deprecated and was removed long ago. Nowadays opus should be used instead.

**Reply by slyoldfox** [reply]
> Hello everybody, I have an intercom/doorbell that uses Flexisip as a SIP server, it's pretty much closed as settings. I saw that it is setup and use the codec:
> 
> ```
> [audio_codec_1]
> mime=speex
> rate=8000
> channels=1
> enabled=1
> ```
> 
> I can authenticate and register into the SIP server and, if I call from a linphone client, I am able to answer and speak. Unfortunately I can't set Baresip to the same codec as the server (speex/8000/1) this is the log I get in Baresip:
> 
> ```
> pi@raspberrypi4:~/.baresip $ baresip 
> baresip v1.0.0 Copyright (C) 2010 - 2020 Alfred E. Heggestad et al.
> Local network address:  IPv4=eth0|192.168.1.28  IPv6=eth0|fdcf:4cbc:9db9:4005:ed44:52a2:6606:ddee
> httpd: listening on 0.0.0.0:8000
> aucodec: opus/48000/2
> aucodec: AMR-WB/16000/1
> aucodec: AMR/8000/1
> aucodec: PCMU/8000/1
> aucodec: PCMA/8000/1
> aufilt: speex_pp
> ausrc: alsa
> auplay: alsa
> medianat: stun
> medianat: turn
> medianat: ice
> mediaenc: srtp
> mediaenc: srtp-mand
> mediaenc: srtp-mandf
> franzoi-outlook.com-RASPI@2123921.bs.iotleg.com: Using media encryption 'srtp'
> Populated 1 account
> Populated 3 contacts
> Populated 5 audio codecs
> Populated 1 audio filter
> Populated 0 video codecs
> Populated 0 video filters
> baresip is ready.
> xxxxxxx-outlook.com-RASPI@xxxxx.xxx.com: {0/TCP/v4} 200 Registration successful (Flexisip/1.0.12 (sofia-sip-nta/2.0)) [1 binding]
> All 1 useragent registered successfully! (105 ms)
> ua: using AF from sdp offer: af=AF_INET
> stream: audio: starting mediaenc 'srtp' (wait_secure=0)
> call: no common audio codecs - rejected
> ```
> 
> Is there a way to be able to set or install this codec (speex) in Baresip? Thanks in advance. Cheers, R.

I've backported the code from 0.5.10 to 2.9.0 and added the necessary files for cmake to do its job.

For people that still need this, you can find the patch on https://gist.github.com/slyoldfox/6c1c97bd43d97ffa428904523d3f63cf

If you'd like this in `main` I'd gladly push this and create a PR @alfredh

**Comment by mrMiimo** [comment]
> The speex codec is deprecated and was removed long ago. Nowadays opus should be used instead.

@cspiel1 Unfortunately I am tied to the firmware of the device and I cannot change the codec, I did some tests with opus but it does not work.

Can you tell me the latest version that contains speex?  and which one can be installed in arm?

**Comment by cspiel1** [comment]
Sorry, no. It is the best you find the latest version working on arm yourself. Then add a patch to your build!

**Comment by mrMiimo** [comment]
thanks Christian @cspiel1 , unfortunately I'm not a programmer. 
I have seen that the last build that contains speex dates back to 0.5.8,
I tried to install 0.5.8 it and it would work but when a call arrive I get the error:

`"Packet is larger than allocated buffer"
Segmentation fault (core dumped) `

and everything crashes..
So I am losing all hope 🤦🏻‍♂️

**Comment by whyteks** [comment, marked as answer]
Hi @mrMiimo 
You don't need to be a programmer to compile programs! Maybe you can just checkout the version you want and maybe, it just compiles. Try it! 

Now, I don't know much about pi, but it's arm, right? Which kind of arm cpu? What "OS" are you running on the pi? 
(you don't have to tell me, these are more questions for yourself..)

If you don't "program", maybe you can "hack", And I wonder if some of these binaries[1] might just run for you.. You might need to manually add dependencies, but it's not programming, it's mainly unpacking archives.

[1] 
http://deb.debian.org/debian/pool/main/b/baresip/baresip-core_0.5.0-4_arm64.deb
http://deb.debian.org/debian/pool/main/b/baresip/baresip-core_0.5.0-4_armel.deb
http://deb.debian.org/debian/pool/main/b/baresip/baresip-core_0.5.0-4_armhf.deb

P.S. I think it would be off-topic to pursue this further on this thread, but good luck!

**Reply by mrMiimo** [reply]
I managed to install version 0.5.7 which in addition to containing speex, also contains mqtt!

I didn't use the packages you listed, but the basic idea was fundamental to succeed!

These days I have been looking for different packages and I was not always able to find the dependencies, but after many tests I succeeded with version 0.5.7.
When I had lost all hope you appeared! thank you @whyteks !!!!!!

**Comment by mrMiimo** [comment]
thanks @whyteks, maybe I explained myself wrong, but with "install" the version 0.5.8 or 0.5.10 (the last one containing the speex codec) I meant to unzip and compile ...ok?
It is not the problem to run the app, much less to configure it, the problem is that when I receive a call, at the time of answering, it crashes with this error:

```
Packet is larger than allocated buffer
Segmentation fault (core dumped)
```
and yes, pi is arm7

**Comment by whyteks** [comment]
OK so that's actually a bug in baresip, (on-topic), related to a need to use speex to communicate with blackbox hardware (valid regardless of arguments about speex being obsoleted or replaced by opus.)

If you want to pursue it, Given that you are happy compiling, you might try a variety of versions. 
git clone...
git checkout v0.5.8 ; compile...
git checkout v0.5.7 ; compile... etc

Of course, if you can run it under gdb and get a backtrace, all the better. 
I really haven't looked at baresip, I never built it, I just saw your post due to coming here to open an issue, so I'm really offering very generic grope-in-the-dark advice. 

Maybe there are some platform dependent buffer sizes defined some place.
Also, if you could, a pcap of the SIP/RTP traffic would be helpful. (not saying that **I** will look at it)

**Reply by mrMiimo** [reply]
thanks, I really appreciate your help!
Tnks so much

**Comment by whyteks** [comment]
BTW, Can you try a packaged version from whatever distro on some x86 hardware? A linux desk/laptop or some such?

**Reply by mrMiimo** [reply]
sometimes i386 packages works, but they weren't always compatible.
The ones that work 100% are armhf.
The version of raspbian that I have chosen does not have the desktop but if you want you can install ...
